// telegraf-broadcast/src/index.js
const { Telegraf, Telegram } = require('telegraf')
const Queue = require('bull')

const getTelegramApi = (bot) => {
    if (bot instanceof Telegraf) {
        return bot.telegram
    }

    if (bot instanceof Telegram) {
        return bot
    }

    throw new Error('bot must be instance of Telegraf or Telegram')
}

const isArrayLike = (item) => {
    return (
        Array.isArray(item) || 
        (
            !!item &&
            typeof item === 'object' &&
            'forEach' in item &&
            typeof item.forEach === 'function' &&
            ('length' in item || 'size' in item)
        )
    )
}

class Broadcaster {
    static queueName = 'tg-broadcast'

    constructor(bot, options) {
        this.options = {
            processes: 1,
            queueName: Broadcaster.queueName,
            bullJobOptions: {},
            bullQueueOptions: {},
            ...options,
        }

        this.telegram = getTelegramApi(bot)

        this.usersProcessed = 0
        this.usersAmount = 0

        this.queue = new Queue(this.queueName, this.options.bullQueueOptions)
        this.queue.process(this.options.processes, this.processor)
    }

    get queueName() {
        return this.options.queueName
    }

    processor = (job, done) => {
        const { chatId, apiMethod, apiData } = job.data
    
        const doneSuccess = (res) => {
            this.usersProcessed += 1
            done(null, res)
        }
    
        const doneError = (err) => {
            this.usersProcessed += 1
            done(err)
        }
    
        if (apiMethod === 'copyMessage') {
            const [fromChatId, messageId, extra] = apiData

            this.telegram.callApi(apiMethod, {
                chat_id: chatId,
                from_chat_id: fromChatId,
                message_id: messageId,
                ...extra,
            })
                .then(doneSuccess)
                .catch(doneError)
        } else {
            this.telegram[apiMethod](chatId, ...apiData)
                .then(doneSuccess)
                .catch(doneError)
        }
    }

    broadcast(chatIds, apiMethod, ...apiData) {
        if (typeof chatIds === 'number' || typeof chatIds === 'string') {
            chatIds = [chatIds]
        }

        if (!isArrayLike(chatIds)) {
            throw new Error('chatIds must be an Array or Array-like of chat/user/channel ids')
        }

        this.usersProcessed = 0
        this.usersAmount = chatIds.length || chatIds.size || 0

        chatIds.forEach(chatId => {
            this.queue.add({ chatId, apiMethod, apiData }, this.options.bullJobOptions)
        })

        return this
    }

    sendMessage(chatIds, fromChatId, messageId, extra) {
        return this.broadcast(chatIds, 'copyMessage',
            fromChatId,
            messageId,
            extra,
        )
    }

    sendText(chatIds, messageText, extra) {
        return this.broadcast(chatIds, 'sendMessage',
            messageText,
            extra,
        )
    }

    sendAudio(chatIds, audio, extra) {
        return this.broadcast(chatIds, 'sendAudio',
            audio,
            extra,
        )
    }

    sendChatAction(chatIds, action) {
        return this.broadcast(chatIds, 'sendChatAction',
            action
        )
    }

    sendDocument(chatIds, document, extra) {
        return this.broadcast(chatIds, 'sendDocument',
            document,
            extra,
        )
    }

    sendGame(chatIds, gameShortName, extra) {
        return this.broadcast(chatIds, 'sendGame',
            gameShortName,
            extra,
        )
    }

    sendLocation(chatIds, latitude, longitude, extra) {
        return this.broadcast(chatIds, 'sendLocation',
            latitude,
            longitude,
            extra,
        )
    }

    sendPhoto(chatIds, photo, extra) {
        return this.broadcast(chatIds, 'sendPhoto',
            photo,
            extra,
        )
    }

    sendMediaGroup(chatIds, media, extra) {
        return this.broadcast(chatIds, 'sendMediaGroup',
            media,
            extra,
        )
    }
    
    sendPoll(chatIds, question, options, extra) {
        return this.broadcast(chatIds, 'sendPoll',
            question,
            options,
            extra,
        )
    }

    sendQuiz(chatIds, question, options, extra) {
        return this.broadcast(chatIds, 'sendQuiz',
            question,
            options,
            extra,
        )
    }

    sendAnimation(chatIds, animation, extra) {
        return this.broadcast(chatIds, 'sendAnimation',
            animation,
            extra,
        )
    }

    sendSticker(chatIds, sticker, extra) {
        return this.broadcast(chatIds, 'sendSticker',
            sticker,
            extra,
        )
    }

    sendVideo(chatIds, video, extra) {
        return this.broadcast(chatIds, 'sendVideo',
            video,
            extra,
        )
    }

    sendVideoNote(chatIds, videoNote, extra) {
        return this.broadcast(chatIds, 'sendVideoNote',
            videoNote,
            extra,
        )
    }

    sendVoice(chatIds, voice, extra) {
        return this.broadcast(chatIds, 'sendVoice',
            voice,
            extra,
        )
    }

    sendDice(chatIds, extra) {
        return this.broadcast(chatIds, 'sendDice',
            extra,
        )
    }

    reset() {
        const queue = this.queue

        this.usersProcessed = 0
        this.usersAmount = 0

        return this.pause()
            .then(() => Promise.all([
                queue.empty(),
                queue.clean(0, 'delayed'),
                queue.clean(0, 'wait'),
                queue.clean(0, 'active'),
                queue.clean(0, 'completed'),
                queue.clean(0, 'failed'),
            ]))
    }

    terminate() {
        const queue = this.queue

        return this.pause()
            .then(() => Promise.all([
                queue.empty(),
                queue.clean(0, 'wait'),
                queue.clean(0, 'active'),
            ]))
    }

    pause() {
        return this.queue.pause()
    }

    resume() {
        return this.queue.resume()
    }

    progress() {
        const percent = ((this.usersProcessed * 100) / this.usersAmount)

        return Number.isNaN(percent) || !Number.isFinite(percent) ? 100 : percent
    }

    onCompleted(callback) {
        return this.queue.on('drained', callback)
    }

    onProcessed(callback) {
        return this.queue.on('completed', callback)
    }

    onFailed(callback) {
        return this.queue.on('failed', callback)
    }

    async failed(formatJob = false) {
        const failedJobs = await this.queue.getFailed()

        return failedJobs.map((job) => formatJob
            ? Broadcaster.formatFailedJob(job)
            : job,
        )
    }

    async status() {
        const queue = this.queue

        const [failedCount, completedCount, activeCount, delayedCount, waitingCount] = await Promise.all([
            queue.getFailedCount(),
            queue.getCompletedCount(),
            queue.getActiveCount(),
            queue.getDelayedCount(),
            queue.getWaitingCount(),
        ])

        return { failedCount, completedCount, activeCount, delayedCount, waitingCount }
    }

    middleware() {
        return Telegraf.tap(ctx => {
            ctx.broadcaster = this
        })
    }

    static formatFailedJob(job) {
        const { failedReason, data } = job

        if (!failedReason) {
            return { data }
        }

        const [code, status, message] = failedReason
            .split(':')
            .map(str => str.trim())

        if (!parseInt(code)) {
            return { data, failedReason }
        }

        return {
            data,
            failedReason: {
                code: parseInt(code),
                status,
                message,
            }
        }
    }
}

module.exports = Broadcaster
